gateway/handler_unixfs_dir.go

```diff
@@ -61,31 +65,31 @@ func (i *handler) serveDirectory(ctx context.Context, w http.ResponseWriter, r *
 
 	// Check if directory has index.html, if so, serveFile
 	idxPath := ipath.Join(contentPath, "index.html")
-	idxResolvedPath, err := i.api.ResolvePath(ctx, idxPath)
-	switch err.(type) {
-	case nil:
-		idx, err := i.api.GetUnixFsNode(ctx, idxResolvedPath)
-		if err != nil {
-			webError(w, err, http.StatusInternalServerError)
-			return false
-		}
-
+	imIndexPath, err := NewImmutablePath(ipath.Join(resolvedPath, "index.html"))
+	if err != nil {
+		webError(w, err, http.StatusInternalServerError)
+		return false
+	}
+	_, idx, err := i.api.Get(ctx, imIndexPath)
+	if err == nil {
 		f, ok := idx.(files.File)
 		if !ok {
-			webError(w, files.ErrNotReader, http.StatusInternalServerError)
+			webError(w, files.ErrNotReader, http.StatusNotFound)
```

Comment:
nit: asking for an existing directory and getting HTTP 404 would be very confusing, let's elaborate error is about index.html inside of it 

```suggestion
			webError(w, fmt.Errorf("%q could not be read: %w", imIndexPath, files.ErrNotReader), http.StatusUnprocessableEntity)
```